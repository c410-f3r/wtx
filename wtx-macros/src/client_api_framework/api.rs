mod attrs;
pub(crate) mod mode;

use crate::{
  client_api_framework::{owned_or_ref::OwnedOrRef, transport_group::TransportGroup},
  misc::{Args, create_ident, parts_from_generics},
};
use proc_macro2::{Ident, Span};
use quote::{ToTokens as _, quote};
use syn::{Item, Path, PathArguments, PathSegment, punctuated::Punctuated, spanned::Spanned as _};

pub(crate) fn api(
  attrs: proc_macro::TokenStream,
  token_stream: proc_macro::TokenStream,
) -> crate::Result<proc_macro::TokenStream> {
  let attr_args: Args = syn::parse(attrs)?;
  let mut item: Item = syn::parse(token_stream)?;

  let attrs::Attrs { error, pkgs_aux, transports, mode: ty } =
    attrs::Attrs::try_from(&attr_args.0)?;

  let pkgs_aux_path = pkgs_aux.as_ref().map_or_else(
    || {
      let mut segments = Punctuated::new();
      segments.push(PathSegment {
        ident: Ident::new("PkgsAux", Span::mixed_site()),
        arguments: PathArguments::None,
      });
      OwnedOrRef::Owned(Path { leading_colon: None, segments })
    },
    OwnedOrRef::Ref,
  );

  let (api_ident, api_generics) = match &mut item {
    Item::Enum(container) => (&mut container.ident, &container.generics),
    Item::Struct(container) => (&mut container.ident, &container.generics),
    Item::Type(container) => (&mut container.ident, &container.generics),
    _ => return Err(crate::Error::NoEnumStructOrType(item.span())),
  };
  let api_string = api_ident.to_string();
  let (api_params, _api_where_predicates) = parts_from_generics(api_generics);

  let api_id_gat_ty = if api_params.is_empty() { quote::quote!() } else { quote!(__ApiParams) };

  let mut buffer = String::new();
  buffer.push_str(&api_string);

  let id_ident = create_ident(&mut buffer, ["Id"]);
  let id_impl = &quote::quote_spanned!(api_ident.span() =>
    #[doc = concat!("Identifier of the `", stringify!(#api_ident), "` API")]
    pub struct #id_ident;

    impl wtx::client_api_framework::ApiId for #id_ident {
      type Api<__ApiParams> = #api_ident<#api_id_gat_ty>;
    }
  );

  let api_impl = if let mode::Mode::Auto = ty {
    Some(quote::quote_spanned!(api_ident.span() =>
      impl<#api_params> wtx::client_api_framework::Api for #api_ident<#api_params> {
        type Error = #error;
        type Id = #id_ident;
      }
    ))
  } else {
    None
  };
  let api_impl_iter = api_impl.into_iter();

  let generic_pair_ident = create_ident(&mut buffer, ["Pair"]);
  let generic_pair_tt = quote::quote_spanned!(api_ident.span() =>
    #[allow(unused_qualifications)]
    #[doc = concat!("[wtx::client_api_framework::misc::Pair] with [", stringify!(#api_ident), "] as the API.")]
    pub type #generic_pair_ident<DRSR, T, TP, #api_params> = wtx::client_api_framework::misc::Pair<
      #pkgs_aux_path<#api_ident<#api_params>, DRSR, TP>,
      T
    >;
  );

  let generic_pkgs_aux_ident = create_ident(&mut buffer, ["PkgsAux"]);
  let generic_pkgs_aux_tt = quote::quote_spanned!(api_ident.span() =>
    #[doc = concat!("[", stringify!(#pkgs_aux_path), "] with an owned [", stringify!(#api_ident), "] as the API.")]
    pub type #generic_pkgs_aux_ident<DRSR, TP, #api_params> = #pkgs_aux_path<#api_ident<#api_params>, DRSR, TP>;
  );

  let generic_mut_pkgs_aux_ident = create_ident(&mut buffer, ["MutPkgsAux"]);
  let generic_mut_pkgs_aux_tt = quote::quote_spanned!(api_ident.span() =>
    #[doc = concat!("[", stringify!(#pkgs_aux_path), "] with a mutable reference of [", stringify!(#api_ident), "] as the API.")]
    pub type #generic_mut_pkgs_aux_ident<'api, DRSR, TP, #api_params> = #pkgs_aux_path<&'api mut #api_ident<#api_params>, DRSR, TP>;
  );

  let mut tys = Vec::new();
  let mut custom_placeholder;

  for transport in transports {
    let [camel_abbr, params] = match transport {
      TransportGroup::Custom(tt) => {
        custom_placeholder = tt.to_string();
        [custom_placeholder.as_str(), custom_placeholder.as_str()]
      }
      TransportGroup::Http => ["Http", "HttpParams"],
      TransportGroup::Stub => ["Stub", "()"],
      TransportGroup::WebSocket => ["Ws", "WsParams"],
    };
    let local_tp_ident = Ident::new(params, Span::mixed_site());
    let local_ty_ident_api_tp = create_ident(&mut buffer, [camel_abbr, "PkgsAux"]);
    let local_ty_ident_api_mut_tp = create_ident(&mut buffer, ["Mut", camel_abbr, "PkgsAux"]);
    buffer.clear();
    let local_ty_ident_tp = create_ident(&mut buffer, [camel_abbr, "PkgsAux"]);
    buffer.push_str(&api_string);
    tys.push(quote::quote!(
      #[allow(unused_qualifications)]
      #[doc = concat!(
        "[", stringify!(#pkgs_aux_path), "] with an owned [",
        stringify!(#api_ident),
        "] as the API and [wtx::client_api_framework::network::",
        stringify!(#local_tp_ident),
        "] as the transport parameter."
      )]
      pub type #local_ty_ident_api_tp<DRSR, #api_params> = #pkgs_aux_path<#api_ident<#api_params>, DRSR, wtx::client_api_framework::network::#local_tp_ident>;

      #[allow(unused_qualifications)]
      #[doc = concat!(
        "[", stringify!(#pkgs_aux_path), "] with a mutable reference of [",
        stringify!(#api_ident),
        "] as the API and [wtx::client_api_framework::network::",
        stringify!(#local_tp_ident),
        "] as the transport parameter."
      )]
      pub type #local_ty_ident_api_mut_tp<'api, DRSR, #api_params> = #pkgs_aux_path<&'api mut #api_ident<#api_params>, DRSR, wtx::client_api_framework::network::#local_tp_ident>;

      #[allow(unused_qualifications)]
      #[doc = concat!(
        "[", stringify!(#pkgs_aux_path), "] with [wtx::client_api_framework::network::",
        stringify!(#local_tp_ident),
        "] as the transport parameter."
      )]
      pub type #local_ty_ident_tp<A, DRSR> = #pkgs_aux_path<A, DRSR, wtx::client_api_framework::network::#local_tp_ident>;
    ));
  }

  let sts_ty = if api_params.is_empty() { quote::quote!(()) } else { api_params.to_token_stream() };

  Ok(
    quote::quote!(
      impl<#api_params> wtx::misc::Lease<#api_ident<#api_params>> for #api_ident<#api_params> {
        #[inline]
        fn lease(&self) -> &#api_ident<#api_params> {
          self
        }
      }

      impl<#api_params> wtx::misc::LeaseMut<#api_ident<#api_params>> for #api_ident<#api_params> {
        #[inline]
        fn lease_mut(&mut self) -> &mut #api_ident<#api_params> {
          self
        }
      }

      impl<#api_params> wtx::misc::SingleTypeStorage for #api_ident<#api_params> {
        type Item = #sts_ty;
      }

      #id_impl
      #(#api_impl_iter)*
      #item
      #generic_pair_tt
      #generic_pkgs_aux_tt
      #generic_mut_pkgs_aux_tt
      #(#tys)*
    )
    .to_token_stream()
    .into(),
  )
}
