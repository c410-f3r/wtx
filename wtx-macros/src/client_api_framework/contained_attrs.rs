use syn::Attribute;

pub(crate) trait ContainedAttrs {
  fn contained_attrs(&mut self) -> Option<&mut Vec<Attribute>>;
}

impl ContainedAttrs for syn::Item {
  fn contained_attrs(&mut self) -> Option<&mut Vec<Attribute>> {
    Some(match *self {
      Self::Const(ref mut item) => item.attrs.as_mut(),
      Self::Enum(ref mut item) => item.attrs.as_mut(),
      Self::ExternCrate(ref mut item) => item.attrs.as_mut(),
      Self::Fn(ref mut item) => item.attrs.as_mut(),
      Self::ForeignMod(ref mut item) => item.attrs.as_mut(),
      Self::Impl(ref mut item) => item.attrs.as_mut(),
      Self::Macro(ref mut item) => item.attrs.as_mut(),
      Self::Mod(ref mut item) => item.attrs.as_mut(),
      Self::Static(ref mut item) => item.attrs.as_mut(),
      Self::Struct(ref mut item) => item.attrs.as_mut(),
      Self::Trait(ref mut item) => item.attrs.as_mut(),
      Self::TraitAlias(ref mut item) => item.attrs.as_mut(),
      Self::Type(ref mut item) => item.attrs.as_mut(),
      Self::Union(ref mut item) => item.attrs.as_mut(),
      Self::Use(ref mut item) => item.attrs.as_mut(),
      _ => return None,
    })
  }
}
