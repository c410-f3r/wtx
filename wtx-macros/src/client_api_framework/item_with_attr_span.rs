use core::mem;
use proc_macro2::{Span, TokenStream};
use syn::Item;

#[derive(Debug)]
pub(crate) struct ItemWithAttrSpan<C, I> {
  pub(crate) _content: C,
  pub(crate) item: I,
  pub(crate) span: Span,
}

impl<'module, C> From<(C, &'module Item, Span)> for ItemWithAttrSpan<C, &'module Item> {
  #[inline]
  fn from((content, item, span): (C, &'module Item, Span)) -> Self {
    Self { _content: content, item, span }
  }
}

impl<'module, C> From<(C, &'module mut Item, Span)> for ItemWithAttrSpan<C, &'module mut Item> {
  #[inline]
  fn from((content, item, span): (C, &'module mut Item, Span)) -> Self {
    Self { _content: content, item, span }
  }
}

impl<'module, C> From<(C, &'module mut Item, Span)> for ItemWithAttrSpan<C, Item> {
  #[inline]
  fn from((content, item, span): (C, &'module mut Item, Span)) -> Self {
    let mut actual_item = Item::Verbatim(TokenStream::new());
    mem::swap(item, &mut actual_item);
    Self { _content: content, item: actual_item, span }
  }
}
