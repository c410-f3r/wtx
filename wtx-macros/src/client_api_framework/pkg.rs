mod data_format;
mod data_format_elems;
mod enum_struct_or_type;
mod fir;
mod keywords;
mod misc;
mod sir;

use crate::{
  client_api_framework::{
    item_with_attr_span::ItemWithAttrSpan,
    pkg::{fir::fir_after_sending_item_values::FirAfterSendingItemValues, misc::unit_type},
  },
  misc::{Args, push_doc},
};
use fir::{
  fir_aux_item_values::FirAuxItemValues,
  fir_before_sending_item_values::FirBeforeSendingItemValues, fir_items_values::FirItemsValues,
  fir_params_items_values::FirParamsItemValues, fir_pkg_attr::FirPkgAttr,
  fir_req_item_values::FirReqItemValues, fir_res_item_values::FirResItemValues,
};
use proc_macro2::{Ident, Span, TokenStream};
use quote::ToTokens as _;
use sir::{sir_final_values::SirFinalValues, sir_pkg_attr::SirPkaAttr};
use syn::{
  Generics, Item, ItemMod, ItemType, Visibility,
  punctuated::Punctuated,
  token::{Eq, Pub, Semi, Type},
};

pub(crate) fn pkg(
  attrs: proc_macro::TokenStream,
  item: proc_macro::TokenStream,
) -> crate::Result<proc_macro::TokenStream> {
  let attr_args: Args = syn::parse(attrs)?;
  let mut item_mod: ItemMod = syn::parse(item)?;
  let items_stub = &mut Vec::new();
  let fiv = FirItemsValues::try_from((
    item_mod.content.as_mut().map_or(items_stub, |el| &mut el.1),
    item_mod.ident.span(),
  ))?;
  let freqdiv = FirReqItemValues::try_from(fiv.req_data)?;
  let mut camel_case_id = {
    let mut string = freqdiv.freqdiv_ident.to_string();
    if let Some(idx) = string.rfind("Req") {
      string.truncate(idx);
    }
    string
  };
  let mut params_item_unit = Item::Verbatim(TokenStream::new());
  let fpiv = if let Some(elem) = fiv.params {
    FirParamsItemValues::try_from(elem)?
  } else {
    params_item_unit = params_item_unit_fn(&mut camel_case_id);
    FirParamsItemValues::try_from(ItemWithAttrSpan {
      _content: (),
      item: &mut params_item_unit,
      span: Span::mixed_site(),
    })?
  };
  let fasiv = fiv.after_sending.map(FirAfterSendingItemValues::try_from).transpose()?;
  let fbsiv = fiv.before_sending.map(FirBeforeSendingItemValues::try_from).transpose()?;
  let faiv = fiv.aux.map(FirAuxItemValues::try_from).transpose()?;
  let fresdiv = FirResItemValues::try_from(fiv.res_data)?;
  let spa = SirPkaAttr::try_from(FirPkgAttr::try_from(&attr_args.0)?)?;
  let SirFinalValues { auxs, package, package_impls } = SirFinalValues::try_from((
    &mut camel_case_id,
    fpiv,
    freqdiv,
    fresdiv,
    spa,
    fasiv,
    faiv,
    fbsiv,
  ))?;
  if let Some(content) = item_mod.content.as_mut() {
    content.1.push(Item::Verbatim(quote::quote!(
      #params_item_unit

      #(#auxs)*
      #package
      #(#package_impls)*
    )));
  }
  Ok(item_mod.into_token_stream().into())
}

fn params_item_unit_fn(camel_case_id: &mut String) -> Item {
  Item::Type(ItemType {
    attrs: {
      let mut attrs = Vec::new();
      push_doc(
        &mut attrs,
        "Corresponding package does not expect any additional custom parameter.",
      );
      attrs
    },
    vis: Visibility::Public(Pub { span: Span::mixed_site() }),
    type_token: Type(Span::mixed_site()),
    ident: {
      let idx = camel_case_id.len();
      camel_case_id.push_str("Params");
      let ident = Ident::new(camel_case_id, Span::mixed_site());
      camel_case_id.truncate(idx);
      ident
    },
    generics: Generics {
      lt_token: None,
      params: Punctuated::new(),
      gt_token: None,
      where_clause: None,
    },
    eq_token: Eq(Span::mixed_site()),
    ty: Box::new(unit_type()),
    semi_token: Semi(Span::mixed_site()),
  })
}
