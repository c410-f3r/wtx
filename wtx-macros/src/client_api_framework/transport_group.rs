use crate::misc::single_elem;
use proc_macro2::{Span, TokenStream};
use quote::ToTokens as _;
use syn::{Meta, Path, PathSegment, Token, punctuated::Punctuated, spanned::Spanned as _};

#[derive(Debug)]
pub(crate) enum TransportGroup {
  Custom(TokenStream),
  Http,
  Stub,
  WebSocket,
}

impl<'attrs> TryFrom<&'attrs Meta> for TransportGroup {
  type Error = crate::Error;

  #[inline]
  fn try_from(from: &'attrs Meta) -> Result<Self, Self::Error> {
    let err = |span: Span| Err(crate::Error::UnknownTransport(span));
    match from {
      Meta::Path(path) => {
        let ps = single_path_segment(path)?;
        Ok(match ps.ident.to_string().as_str() {
          "http" => Self::Http,
          "stub" => Self::Stub,
          "ws" => Self::WebSocket,
          _ => return err(ps.ident.span()),
        })
      }
      Meta::List(meta_list) => {
        let ps = single_path_segment(&meta_list.path)?;
        if ps.ident == "custom" {
          let metas = meta_list.parse_args_with(Punctuated::parse_terminated)?;
          if let Meta::Path(path) = single_nested(&metas)? {
            return Ok(Self::Custom(path.to_token_stream()));
          }
        }
        err(ps.ident.span())
      }
      Meta::NameValue(mnv) => err(mnv.span()),
    }
  }
}

fn single_nested(nested: &Punctuated<Meta, Token![,]>) -> crate::Result<&Meta> {
  single_elem(nested.iter()).ok_or_else(|| crate::Error::UnknownTransport(Span::mixed_site()))
}

fn single_path_segment(path: &Path) -> crate::Result<&PathSegment> {
  single_elem(path.segments.iter())
    .ok_or_else(|| crate::Error::UnknownTransport(Span::mixed_site()))
}
