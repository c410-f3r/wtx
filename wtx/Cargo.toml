[dependencies]
aes-gcm = { default-features = false, optional = true, version = "0.11.0-pre.2" }
arbitrary = { default-features = false, features = ["derive_arbitrary"], optional = true, version = "1.0" }
argon2 = { default-features = false, optional = true, version = "0.5" }
base64 = { default-features = false, features = ["alloc"], optional = true, version = "0.22" }
borsh = { default-features = false, features = ["derive"], optional = true, version = "1.0" }
chrono = { default-features = false, optional = true, version = "0.4" }
cl-aux = { default-features = false, optional = true, features = ["alloc"], version = "5.0" }
crypto-common = { default-features = false, optional = true, version = "0.1" }
digest = { default-features = false, features = ["mac"], optional = true, version = "0.10" }
fastrand = { default-features = false, optional = true, version = "2.0" }
flate2 = { default-features = false, features = ["zlib-rs"], optional = true, version = "1.0" }
foldhash = { default-features = false, optional = true, version = "0.1" }
hashbrown = { default-features = false, features = ["inline-more"], optional = true, version = "0.15" }
hmac = { default-features = false, optional = true, version = "0.12" }
httparse = { default-features = false, optional = true, version = "1.0" }
matchit = { default-features = false, optional = true, version = "0.8" }
memchr = { default-features = false, optional = true, version = "2.0" }
quick-protobuf = { default-features = false, optional = true, version = "0.8" }
rand_chacha = { default-features = false, optional = true, version = "0.3" }
rand_core = { default-features = false, optional = true, version = "0.6" }
ring = { default-features = false, optional = true, version = "0.17" }
rust_decimal = { default-features = false, features = ["maths"], optional = true, version = "1.0" }
rustls-pemfile = { default-features = false, optional = true, version = "2.0" }
rustls-pki-types = { default-features = false, optional = true, version = "1.0" }
serde = { default-features = false, features = ["alloc", "derive"], optional = true, version = "1.0" }
serde_json = { default-features = false, features = ["alloc"], optional = true, version = "1.0" }
sha1 = { default-features = false, optional = true, version = "0.10" }
sha2 = { default-features = false, optional = true, version = "0.10" }
simdutf8 = { default-features = false, features = ["aarch64_neon"], optional = true, version = "0.1" }
tokio = { default-features = false, features = ["io-util", "net", "rt", "sync", "time"], optional = true, version = "1.0" }
tokio-rustls = { default-features = false, features = ["ring"], optional = true, version = "0.26" }
tracing = { default-features = false, features = ["attributes"], optional = true, version = "0.1" }
tracing-subscriber = { default-features = false, features = ["env-filter", "fmt"], optional = true, version = "0.3" }
tracing-tree = { default-features = false, optional = true, version = "0.4" }
webpki-roots = { default-features = false, optional = true, version = "0.26" }
x509-certificate = { default-features = false, optional = true, version = "0.23" }

[features]
arbitrary = ["dep:arbitrary", "std"]
base64 = ["dep:base64"]
borsh = ["dep:borsh", "std"]
chrono = ["dep:chrono"]
cl-aux = ["dep:cl-aux"]
client-api-framework = ["data-transformation"]
crypto-common = ["dep:crypto-common"]
data-transformation = []
database = []
default = []
digest = ["dep:digest"]
executor = []
fastrand = ["dep:fastrand"]
flate2 = ["dep:flate2"]
foldhash = ["dep:foldhash", "hashbrown?/default-hasher"]
grpc = ["data-transformation", "http-client-framework", "http-server-framework"]
hashbrown = ["dep:hashbrown"]
hmac = ["dep:hmac"]
http = []
http-client-framework = ["http2", "pool", "std"]
http-cookie = ["chrono/alloc", "http"]
http-cookie-secure = ["aes-gcm/aes", "aes-gcm/alloc", "base64", "digest", "http-cookie"]
http-server-framework = ["http2", "tokio"]
http-session = ["chrono/serde", "http-cookie-secure", "serde_json"]
http2 = ["foldhash", "hashbrown", "http"]
httparse = ["dep:httparse"]
memchr = ["dep:memchr"]
nightly = ["hashbrown?/nightly"]
optimization = ["memchr", "simdutf8"]
pool = []
postgres = ["base64", "crypto-common", "database", "digest", "foldhash", "hashbrown", "hmac", "sha2"]
quick-protobuf = ["dep:quick-protobuf", "std"]
rand_chacha = ["dep:rand_chacha", "dep:rand_core"]
ring = ["dep:ring"]
rust_decimal = ["dep:rust_decimal"]
schema-manager = ["database", "chrono"]
schema-manager-dev = ["schema-manager"]
serde = ["cl-aux?/serde", "dep:serde"]
serde_json = ["serde", "dep:serde_json", "std"]
sha1 = ["dep:sha1"]
sha2 = ["dep:sha2"]
simdutf8 = ["dep:simdutf8"]
std = ["aes-gcm?/std", "argon2?/std", "base64?/std", "borsh?/std", "chrono?/std", "cl-aux?/std", "crypto-common?/std", "digest?/std", "fastrand?/std", "foldhash?/std", "hmac?/std", "httparse?/std", "memchr?/std", "quick-protobuf?/std", "rand_chacha?/std", "rand_core?/std", "ring?/std", "rust_decimal?/std", "rustls-pemfile?/std", "rustls-pki-types?/std", "serde?/std", "serde_json?/std", "sha1?/std", "sha2?/std", "simdutf8?/std", "tracing?/std", "tracing-subscriber?/std"]
tokio = ["std", "dep:tokio"]
tokio-rustls = ["ring", "dep:rustls-pemfile", "dep:rustls-pki-types", "tokio", "dep:tokio-rustls"]
tracing = ["dep:tracing"]
tracing-subscriber = ["dep:tracing-subscriber"]
web-socket = ["http"]
web-socket-handshake = ["base64", "httparse", "sha1", "web-socket"]
webpki-roots = ["dep:webpki-roots"]
x509-certificate = ["dep:x509-certificate"]

_async-tests = ["tokio/macros", "tokio/net", "tokio/rt-multi-thread", "tokio/time"]
_bench = []
_integration-tests = ["serde_json?/raw_value"]
_tracing-tree = ["tracing", "tracing-subscriber", "dep:tracing-tree"]

[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(kani)'] }

[package]
authors = ["Caio Fernandes <c410.f3r@gmail.com>"]
categories = ["asynchronous", "database", "network-programming", "no-std", "web-programming"]
description = "A collection of different transport implementations and related tools focused primarily on web technologies."
documentation = "https://docs.rs/wtx"
edition = "2021"
exclude = ["tests"]
homepage = "https://c410-f3r.github.io/wtx"
keywords = ["api", "database", "http", "network", "websocket"]
license = "Apache-2.0"
name = "wtx"
readme = "README.md"
repository = "https://github.com/c410-f3r/wtx"
rust-version = "1.82"
version = "0.23.0"

[package.metadata.docs.rs]
all-features = true
