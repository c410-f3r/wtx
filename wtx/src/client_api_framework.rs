//! A flexible client API framework for writing asynchronous, fast, organizable, scalable and maintainable applications.

#[macro_use]
mod macros;

mod api;
mod client_api_framework_error;
pub mod misc;
pub mod network;
pub mod pkg;
mod send_bytes_ty;
#[cfg(test)]
mod tests;

pub use api::{Api, ApiId};
pub use client_api_framework_error::ClientApiFrameworkError;
pub use send_bytes_ty::SendBytesSource;
