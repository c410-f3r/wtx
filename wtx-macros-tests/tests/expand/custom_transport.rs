//! Custom transport through `transport(Custom)`.

use wtx::client_api_framework::{
  network::{
    transport::{ReceivingTransport, SendingTransport, Transport, TransportParams},
    TransportGroup,
  },
  pkg::{Package, PkgsAux},
  Api, SendBytesSource
};

struct CustomTransport;

impl ReceivingTransport<CustomTransportParams> for CustomTransport {
  #[inline]
  async fn recv<A, DRSR>(
    &mut self,
    _: &mut PkgsAux<A, DRSR, CustomTransportParams>,
  ) -> Result<Range<usize>, A::Error>
  where
    A: Api,
  {
    Ok(())
  }
}

impl SendingTransport<CustomTransportParams> for CustomTransport {
  async fn send_bytes<A, DRSR>(
    &mut self,
    _: SendBytesSource<'_>,
    _: &mut PkgsAux<A, DRSR, CustomTransportParams>,
  ) -> Result<(), A::Error>
  where
    A: Api
  {
    Ok(())
  }

  async fn send_pkg<A, DRSR, P>(
    &mut self,
    _: &mut P,
    _: &mut PkgsAux<A, DRSR, CustomTransportParams>,
  ) -> Result<(), A::Error>
  where
    A: Api,
    P: Package<A, DRSR, Self, CustomTransportParams>,
  {
    Ok(())
  }
}

impl Transport<CustomTransportParams> for CustomTransport {
  const GROUP: TransportGroup = TransportGroup::Custom("Custom");
  type Inner = Self;
  type ReqId = ();
}

struct CustomTransportParams((), ());

impl TransportParams for CustomTransportParams {
  type ExternalRequestParams = ();
  type ExternalResponseParams = ();

  fn ext_params(&self) -> (&Self::ExternalRequestParams, &Self::ExternalResponseParams) {
    (&self.0, &self.1)
  }

  fn ext_params_mut(&mut self) -> (&mut Self::ExternalRequestParams, &mut Self::ExternalResponseParams) {
    (&mut self.0, &mut self.1)
  }

  fn reset(&mut self) {}
}

type Nothing = ();

#[wtx_macros::pkg(data_format(json), id(super::Nothing), transport(custom(crate::CustomTransportParams)))]
mod pkg {
  #[pkg::req_data]
  struct Req;

  #[pkg::res_data]
  struct Res;
}

fn main() {}