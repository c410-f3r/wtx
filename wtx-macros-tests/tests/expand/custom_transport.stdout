#![feature(prelude_import)]
//! Custom transport through `transport(Custom)`.
#[macro_use]
extern crate std;
#[prelude_import]
use std::prelude::rust_2021::*;

use wtx::client_api_framework::{
    network::{
        transport::{
            ReceivingTransport, SendingTransport, Transport, TransportParams,
        },
        TransportGroup,
    },
    pkg::{Package, PkgsAux},
    Api, SendBytesSource,
};

struct CustomTransport;

impl ReceivingTransport<CustomTransportParams> for CustomTransport {
    #[inline]
    async fn recv<A,
        DRSR>(&mut self, _: &mut PkgsAux<A, DRSR, CustomTransportParams>,
        _: Self::ReqId) -> Result<(), A::Error> where A: Api {
        Ok(())
    }
}

impl SendingTransport<CustomTransportParams> for CustomTransport {
    async fn send_bytes<A,
        DRSR>(&mut self, _: SendBytesSource<'_>,
        _: &mut PkgsAux<A, DRSR, CustomTransportParams>)
        -> Result<(), A::Error> where A: Api {
        Ok(())
    }

    async fn send_pkg<A, DRSR,
        P>(&mut self, _: &mut P,
        _: &mut PkgsAux<A, DRSR, CustomTransportParams>)
        -> Result<(), A::Error> where A: Api,
        P: Package<A, DRSR, Self, CustomTransportParams> {
        Ok(())
    }
}

impl Transport<CustomTransportParams> for CustomTransport {
    const GROUP: TransportGroup = TransportGroup::Custom("Custom");
    type Inner = Self;
    type ReqId = ();
}

struct CustomTransportParams((), ());

impl TransportParams for CustomTransportParams {
    type ExternalRequestParams = ();
    type ExternalResponseParams = ();

    fn ext_params(&self)
        -> (&Self::ExternalRequestParams, &Self::ExternalResponseParams) {
        (&self.0, &self.1)
    }

    fn ext_params_mut(&mut self)
        ->
            (&mut Self::ExternalRequestParams,
            &mut Self::ExternalResponseParams) {
        (&mut self.0, &mut self.1)
    }

    fn reset(&mut self) {}
}

type Nothing = ();

mod pkg {
    #[doc = "Expected data request that will be sent to the server."]
    struct Req;

    #[doc = "Expected data response returned by the server."]
    struct Res;
    #[doc =
    "Corresponding package does not expect any additional custom parameter."]
    pub type Params = ();
    #[doc =
    r" Package containing all the expected parameters and data necessary to manage and issue"]
    #[doc = r" a request."]
    #[doc = r""]
    #[doc =
    r" For more information, please see the official API's documentation."]
    pub struct Pkg<C> where  {
        #[doc = r" Content. Data format containing request data."]
        pub content: C,
        #[doc = r" Parameters. Used across the package lifetime."]
        pub params: Params<>,
    }
    #[automatically_derived]
    impl<C: ::core::fmt::Debug> ::core::fmt::Debug for Pkg<C> {
        #[inline]
        fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
            ::core::fmt::Formatter::debug_struct_field2_finish(f, "Pkg",
                "content", &self.content, "params", &&self.params)
        }
    }
    impl<__API, __API_PARAMS, __DRSR, __TRANSPORT>
        wtx::client_api_framework::pkg::Package<__API, __DRSR, __TRANSPORT,
        crate::CustomTransportParams> for
        Pkg<wtx::de::protocol::VerbatimEncoder<Req<>>> where
        wtx::de::protocol::VerbatimEncoder<Req<>>: wtx::de::Encode<wtx::de::format::De<__DRSR>>,
        for<'__de> wtx::de::protocol::VerbatimDecoder<Res<>>: wtx::de::DecodeSeq<'__de,
        wtx::de::format::De<__DRSR>>,
        __API: wtx::client_api_framework::Api<Error =
        <<super::Nothing as
        wtx::client_api_framework::ApiId>::Api<__API_PARAMS> as
        wtx::client_api_framework::Api>::Error, Id = super::Nothing> +
        wtx::misc::LeaseMut<<super::Nothing as
        wtx::client_api_framework::ApiId>::Api<__API_PARAMS>> +
        wtx::misc::SingleTypeStorage<Item = __API_PARAMS> {
        type ExternalRequestContent =
            wtx::de::protocol::VerbatimEncoder<Req<>>;
        type ExternalResponseContent<'__de> =
            wtx::de::protocol::VerbatimDecoder<Res<>>;
        type PackageParams = Params<>;
        #[inline]
        async fn after_sending(&mut self,
            (_api, _bytes, _drsr):
                (&mut __API, &mut wtx::collection::Vector<u8>, &mut __DRSR),
            (_trans, _trans_params):
                (&mut __TRANSPORT, &mut crate::CustomTransportParams))
            -> Result<(), __API::Error> {
            Ok(())
        }
        #[inline]
        async fn before_sending(&mut self,
            (_api, _bytes, _drsr):
                (&mut __API, &mut wtx::collection::Vector<u8>, &mut __DRSR),
            (_trans, _trans_params):
                (&mut __TRANSPORT, &mut crate::CustomTransportParams))
            -> Result<(), __API::Error> {
            Ok(())
        }
        #[inline]
        fn ext_req_content(&self) -> &Self::ExternalRequestContent {
            &self.content
        }
        #[inline]
        fn ext_req_content_mut(&mut self)
            -> &mut Self::ExternalRequestContent {
            &mut self.content
        }
        #[inline]
        fn pkg_params(&self) -> &Self::PackageParams { &self.params }
        #[inline]
        fn pkg_params_mut(&mut self) -> &mut Self::PackageParams {
            &mut self.params
        }
    }
    impl<'__is_mut, __API, __API_PARAMS, __DRSR, __TRANSPORT>
        wtx::client_api_framework::pkg::Package<__API, __DRSR, __TRANSPORT,
        &'__is_mut mut crate::CustomTransportParams> for
        Pkg<wtx::de::protocol::VerbatimEncoder<Req<>>> where
        wtx::de::protocol::VerbatimEncoder<Req<>>: wtx::de::Encode<wtx::de::format::De<__DRSR>>,
        for<'__de> wtx::de::protocol::VerbatimDecoder<Res<>>: wtx::de::DecodeSeq<'__de,
        wtx::de::format::De<__DRSR>>,
        __API: wtx::client_api_framework::Api<Error =
        <<super::Nothing as
        wtx::client_api_framework::ApiId>::Api<__API_PARAMS> as
        wtx::client_api_framework::Api>::Error, Id = super::Nothing> +
        wtx::misc::LeaseMut<<super::Nothing as
        wtx::client_api_framework::ApiId>::Api<__API_PARAMS>> +
        wtx::misc::SingleTypeStorage<Item = __API_PARAMS> {
        type ExternalRequestContent =
            wtx::de::protocol::VerbatimEncoder<Req<>>;
        type ExternalResponseContent<'__de> =
            wtx::de::protocol::VerbatimDecoder<Res<>>;
        type PackageParams = Params<>;
        #[inline]
        async fn after_sending(&mut self,
            (_api, _bytes, _drsr):
                (&mut __API, &mut wtx::collection::Vector<u8>, &mut __DRSR),
            (_trans, _trans_params):
                (&mut __TRANSPORT,
                &mut &'__is_mut mut crate::CustomTransportParams))
            -> Result<(), __API::Error> {
            Ok(())
        }
        #[inline]
        async fn before_sending(&mut self,
            (_api, _bytes, _drsr):
                (&mut __API, &mut wtx::collection::Vector<u8>, &mut __DRSR),
            (_trans, _trans_params):
                (&mut __TRANSPORT,
                &mut &'__is_mut mut crate::CustomTransportParams))
            -> Result<(), __API::Error> {
            Ok(())
        }
        #[inline]
        fn ext_req_content(&self) -> &Self::ExternalRequestContent {
            &self.content
        }
        #[inline]
        fn ext_req_content_mut(&mut self)
            -> &mut Self::ExternalRequestContent {
            &mut self.content
        }
        #[inline]
        fn pkg_params(&self) -> &Self::PackageParams { &self.params }
        #[inline]
        fn pkg_params_mut(&mut self) -> &mut Self::PackageParams {
            &mut self.params
        }
    }
}

fn main() {}
